File: Python/Decode Ways.py
class Solution:
    # @param s, a string
    # @return an integer
    def numDecodings(self, s):
		if s=="" or s[0]=="0":return 0
		dp=[1,1]
		for i in range(2,len(s)+1):
			if 10<=int(s[i-2:i])<=26 and 0<int(s[i-1])<10:
				dp.append(dp[i-2]+dp[i-1])
			elif 10<=int(s[i-2:i])<=26:
				dp.append(dp[i-2])
			elif 0<int(s[i-1])<10:
				dp.append(dp[i-1])
			else:
				return 0
		return dp[len(s)]

File: Java/Decode Ways.java
class Decode Ways{
	public int numDecodings(String s) {
        int n = s.length();
        if (n==0) return 0;
        int[] dp = new int[n+1];
        dp[0] = 1;
        if (isValid(s.substring(0,1))) dp[1] = 1;
        else dp[1] = 0;
        for(int i=2; i<=n;i++){
            if (isValid(s.substring(i-1,i)))
                dp[i] = dp[i-1];
            if (isValid(s.substring(i-2,i)))
                dp[i] += dp[i-2];
        }
        return dp[n];
    }
    
    public boolean isValid(String s){
        if (s.charAt(0)=='0') return false;
        int code = Integer.parseInt(s);
        return code>=1 && code<=26;
    }
}